File: scripts/validate.mjs
import { access, readFile } from "node:fs/promises";
import { execFileSync } from "node:child_process";
import path from "node:path";
import process from "node:process";

const root = path.resolve(new URL("..", import.meta.url).pathname);
const required = [
  "site/index.html", "site/config.js", "site/.nojekyll", "site/assets/app.js", "site/assets/styles.css",
  "site/templates/social-signals-template.csv", "worker/wrangler.jsonc", "worker/src/index.js",
  "worker/migrations/0001_initial.sql", "worker/migrations/0002_social_signals.sql",
  ".github/workflows/pages.yml", ".github/workflows/worker.yml", "README.md", "QUICK_START.md"
];

for (const file of required) await access(path.join(root, file));
for (const file of [
  "site/assets/app.js", "site/service-worker.js", "worker/src/index.js",
  "worker/src/lib/utils.js", "worker/src/lib/scoring.js", "worker/src/lib/db.js",
  "worker/src/lib/food-taxonomy.js", "worker/src/collectors/google-trends.js", "worker/src/collectors/youtube.js"
]) {
  execFileSync(process.execPath, ["--check", path.join(root, file)], { stdio: "inherit" });
}
const html = await readFile(path.join(root, "site/index.html"), "utf8");
for (const reference of ["./assets/styles.css", "./assets/app.js", "./config.js", "./assets/icon.svg", "socialSignalsBody", "socialAdminToken"]) {
  if (!html.includes(reference)) throw new Error(`Missing HTML reference or element: ${reference}`);
}
console.log(`Validated ${required.length} required files, social inbox elements, and JavaScript syntax.`);

File: scripts/validate.py
#!/usr/bin/env python3
"""Dependency-free static validation for the GitHub Pages package."""
from __future__ import annotations

import json
from html.parser import HTMLParser
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]


class AssetParser(HTMLParser):
    def __init__(self) -> None:
        super().__init__()
        self.assets: list[str] = []
        self.ids: set[str] = set()

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        values = dict(attrs)
        if values.get("id"):
            if values["id"] in self.ids:
                raise AssertionError(f"Duplicate HTML id: {values['id']}")
            self.ids.add(values["id"])
        if tag == "script" and values.get("src"):
            self.assets.append(values["src"])
        if tag == "link" and values.get("href"):
            self.assets.append(values["href"])


def validate_json(relative: str) -> None:
    with (ROOT / relative).open(encoding="utf-8") as handle:
        json.load(handle)


def validate_html(relative: str) -> None:
    parser = AssetParser()
    parser.feed((ROOT / relative).read_text(encoding="utf-8"))
    for asset in parser.assets:
        if asset.startswith(("http://", "https://", "#", "data:")):
            continue
        path = ROOT / asset.removeprefix("./")
        assert path.exists(), f"Missing asset referenced by {relative}: {asset}"


def main() -> None:
    for relative in ["data/seed.json", "manifest.webmanifest", "apps-script/appsscript.json"]:
        validate_json(relative)
    for relative in ["index.html", "404.html"]:
        validate_html(relative)
    required = [
        ".nojekyll", "config.js", "service-worker.js", "assets/app.js", "assets/styles.css",
        ".github/workflows/pages.yml", "apps-script/Code.gs", "README.md"
    ]
    for relative in required:
        assert (ROOT / relative).exists(), f"Required file missing: {relative}"
    seed = json.loads((ROOT / "data/seed.json").read_text(encoding="utf-8"))
    trend_ids = {row["id"] for row in seed["trends"]}
    assert trend_ids, "Seed must contain at least one trend"
    assert all(row["trend_id"] in trend_ids for row in seed["observations"]), "Orphan observation in seed"
    assert all(row["trend_id"] in trend_ids for row in seed["experiments"]), "Orphan experiment in seed"
    print("Static package validation passed.")


if __name__ == "__main__":
    main()
